File: 2021/08/14/summary_ranges.py
"""
Problem: https://leetcode.com/problems/summary-ranges/
Author: Andrew Jarombek
Date: 8/14/2021
"""

from typing import List


def summary_ranges(nums: List[int]) -> List[str]:
    if len(nums) == 0:
        return []

    result = []

    start = nums[0]
    last = None

    for num in nums[1:]:
        prev = start if last is None else last

        if prev + 1 != num:
            if last is not None:
                result.append(f"{start}->{last}")
            else:
                result.append(f"{start}")

            start = num
            last = None
        else:
            last = num

    if start is not None:
        if last is not None:
            result.append(f"{start}->{last}")
        else:
            result.append(f"{start}")

    return result


if __name__ == '__main__':
    assert summary_ranges([]) == []
    assert summary_ranges([1]) == ["1"]
    assert summary_ranges([0, 2, 3, 4, 6, 8, 9]) == ["0", "2->4", "6", "8->9"]

File: 2022/02/19/summaryRanges.ts
/**
 * https://leetcode.com/problems/summary-ranges/
 * @author Andrew Jarombek
 * @since 2/19/2022
 */

function summaryRanges(nums: number[]): string[] {
    if (!nums) {
        return [];
    }

    nums.push(nums[-1]);
    let first = nums[0];
    let prev = nums[0];
    let result = [];

    for (let index = 1; index < nums.length; index++) {
        const num = nums[index];

        if (num == prev + 1) {
            prev += 1;
        } else {
            if (prev == first) {
                result.push(`${first}`);
            } else {
                result.push(`${first}->${prev}`);
            }

            first = num;
            prev = num;
        }
    }

    return result;
}